### ISSUE TITLE
Feature request: Dynamic query total/free GPU memory from within Chapel

### ISSUE THREAD

**Issue by psath** [chapel-lang/chapel#23681]
I waffled about whether this should be a separate issue from #23638 which seems to deal with _static_ properties of GPUs, whereas this is about a more _dynamic_ property. If you all feel they belong together, by all means close and integrate with that issue.

I've got a use-case where I want to oversubscribe a GPU til the total working-set size across streams is just under the total GPU memory. I can implement most of that in user code with an `atomic` tracking the on-GPU size of my pending/in-flight tasks and CAS to in/decrement as I create new pending / finish old outstanding tasks. (I understand this is still subject to races if any other program uses the GPU, or the runtime needs to grab additional that I am not manually tracking.)

However, I do need to be able to query the total GPU memory as well as the Chapel runtime's "fixed" post-init usage (relevant to #23466). Both of those could be achieved by wrapping `cuda/hipMemGetInfo`.

In the immediate term I could fudge these by exposing some top-level `config var`s and trust the user to specify the right numbers with a little leeway for under-estimation, but it would be cleaner to do it within the GPU sub-locale. Alternatively, if I were to use a C-interop-wrapped version of the native function within an `on gpu[x]`, I assume that would be run on the `x-th` GPU? 

There might be other dynamic queries that would make sense to expose via Chapel, so feel free to extend this as an umbrella issue to them, but none are jumping out at me.

**Comment by e-kayrakli** [comment]
This one's interesting. We do support `MemDiagnostics` inside GPU locales. So, what I have below comes close to what you're asking for:

```chpl
use MemDiagnostics;

config const arrs = 10;
config const n = 100;
config const verboseMem = false;

if verboseMem then startVerboseMem();
on here.gpus[0] {
  writeln("Before array alloc: ", memoryUsed());
  var A: [1..n] int;
  writeln("After array alloc: ", memoryUsed());
  writeln(A[1]);
  var B: [1..n] int;
  writeln("After array alloc: ", memoryUsed());
  writeln(B[1]);
}
if verboseMem then stopVerboseMem();
```

Running this with `--memTrack` flag, you'll get:

```
Before array alloc: 0
After array alloc: 67176
0
After array alloc: 68232
0
```

The first number is huge, and threw me off, that's why I added calls to verbose memory reporting as well. So, if you run with `--memTrack --verboseMem`, you'll get:

```
0 (gpu 0): --:0: allocate 64B of io buffer or bytes at 0x7f10a88b11c0
0 (gpu 0): --:0: allocate 480B of io buffer or bytes at 0x7f10a894b000
0 (gpu 0): --:175: allocate 40B of io buffer or bytes at 0x7f10a88110f0
0 (gpu 0): --:156: allocate 65536B of io buffer or bytes at 0x7f10a894d000
Before array alloc: 0
0 (gpu 0): $CHPL_HOME/modules/internal/DefaultRectangular.chpl:103: allocate 88B of DefaultRectangularDom(1,int(64),one) at 0x7f1066e00600
0 (gpu 0): $CHPL_HOME/modules/internal/DefaultRectangular.chpl:699: allocate 168B of DefaultRectangularArr(1,int(64),one,int(64),int(64)) at 0x7f1066e00800
0 (gpu 0): $CHPL_HOME/modules/internal/ChapelBase.chpl:1604: allocate 800B of array elements at 0x7f1067000200
0 (gpu 0): --:0: allocate 8B of gpu layer utility space at 0x7f10a891d050
After array alloc: 67176
0
0 (gpu 0): $CHPL_HOME/modules/internal/DefaultRectangular.chpl:103: allocate 88B of DefaultRectangularDom(1,int(64),one) at 0x7f1066e00a00
0 (gpu 0): $CHPL_HOME/modules/internal/DefaultRectangular.chpl:699: allocate 168B of DefaultRectangularArr(1,int(64),one,int(64),int(64)) at 0x7f1066e00c00
0 (gpu 0): $CHPL_HOME/modules/internal/ChapelBase.chpl:1604: allocate 800B of array elements at 0x7f1067000600
After array alloc: 68232
0
0 (gpu 0): $CHPL_HOME/modules/internal/DefaultRectangular.chpl:1176: free 800B of array elements at 0x7f1067000600
0 (gpu 0): $CHPL_HOME/modules/internal/ChapelBase.chpl:2598: free 168B of DefaultRectangularArr(1,int(64),one,int(64),int(64)) at 0x7f1066e00c00
0 (gpu 0): $CHPL_HOME/modules/internal/ChapelBase.chpl:2598: free 88B of DefaultRectangularDom(1,int(64),one) at 0x7f1066e00a00
0 (gpu 0): $CHPL_HOME/modules/internal/DefaultRectangular.chpl:1176: free 800B of array elements at 0x7f1067000200
0 (gpu 0): $CHPL_HOME/modules/internal/ChapelBase.chpl:2598: free 168B of DefaultRectangularArr(1,int(64),one,int(64),int(64)) at 0x7f1066e00800
0 (gpu 0): $CHPL_HOME/modules/internal/ChapelBase.chpl:2598: free 88B of DefaultRectangularDom(1,int(64),one) at 0x7f1066e00600
```

So that big jump is from IO buffers. And allocating a 100-element `int(64)` array costs you 800 bytes and then some for array metadata. I think something like this can help to some extent, though it doesn't cover everything you're asking for.

All what I have here relies on our runtime's internal memory tracking. So it misses the memory held by the GPU driver, which can be sizeable and dynamic by itself I think. Do we allow our memtracking support to use the device queries which can be more realistic, but then potentially creating a discrepancy with verbose memory output?

An alternative is to have something in the `locale` type that can report available memory similar to `locale.physicalMemory`.

**Comment by psath** [comment]
Iiiiinteresting. MemDiagnostics is new to me, but happy to hear it works inside the GPU locale!

I can probably cobble together a working short-term implementation with the info it provides, external magic numbers from `nvidia-smi` (total physical and driver) and and appropriately-conservative upper% threshold to leave room for any dynamic needs outside of Chapel-space... It might take a bit, but if I can capture a small pattern of that I'll add it to the discussion.

**Comment by e-kayrakli** [comment]
Here's a quick variation that I whipped up. This one adds `availableMemory` on the `locale` type as a tertiary method. Internals has some non-user-facing stuff. I haven't tested this with hip, nor checked the doc. I am just assuming that the interface is identical.

```chpl
use MemDiagnostics;

proc locale.availableMemory {
  use ChplConfig;
  if CHPL_LOCALE_MODEL == "flat" {
    compilerError("Only supported with GPU locale model");
  }
  else if ((this._instance):GPULocale) == nil {
    writeln("Warning: locale.availableMemory always returns 0 in host for now");
  }
  else {
    param procName = if CHPL_GPU == "nvidia" then "cuMemGetInfo"
                                             else "hipMemGetInfo";
    extern procName proc getInfo(ref free, ref total);
    var free, total: int;
    getInfo(free, total);
    return (free, total);
  }
  return (0,0);
}

config const arrs = 10;
config const n = 100;
config const verboseMem = false;

if verboseMem then startVerboseMem();
on here.gpus[0] {
  writeln("(free,total) memory before alloc: ", here.availableMemory);
  var A: [1..n] int;
  writeln("(free,total) memory after alloc: ", here.availableMemory);
  writeln(A[1]);
  var B: [1..n] int;
  writeln("(free,total) memory after alloc: ", here.availableMemory);
  writeln(B[1]);
}
if verboseMem then stopVerboseMem();
```

An interesting observation from this is that the "Free" memory doesn't change between the array allocations by default. You see it change if you set `--n=1_000_000` or something large. Probably CUDA reports/track on the granularity of pages so small allocations are "lost" in terms of tracking. There's a philosophical discussion of what's actually "free" here. As our baked-in memtracking seems more precise on the surface as it tracks based on the user's requested allocation amounts. But if subsequent allocations indeed can't allocate from pages that have been partially used before, is the memory in such pages really "free"?

TL;DR: I feel stronger that we should provide interface for both ways (1 is already there through MemDiagnostics) because the values reported by each can have different cases where they can be useful.
